Core/Application/CQRS/SoilLayerData/InsertSoilLayerTableCommand.cs

```diff
@@ -42,48 +42,65 @@ private Unit Handler(InsertSoilLayerTableCommand request)
         {
             var traits = _context.GetTraitsFromColumns(request.Table, request.Skip, request.Type);
 
-            SoilLayerData data;
-
-            foreach (DataRow row in request.Table.Rows)
+            IEnumerable<SoilLayerData> convertRow(DataRow row)
             {
                 var all = _context.Plots
                         .Where(p => p.Treatment.Experiment.Name == row[0].ToString());
 
                 if (row[1].ToString().ToLower() != "all")
                     all = all.Where(p => p.Column == Convert.ToInt32(row[1]));
 
+                request.IncrementProgress();
+
                 foreach (var plot in all)
                 {
                     for (int i = 5; i < row.ItemArray.Length; i++)
                     {
                         if (row[i] is DBNull || row[i] is "") continue;
 
                         var value = Convert.ToDouble(row[i]);
-                        data = new SoilLayerData()
+                        yield return new SoilLayerData()
```

Comment:
what is this actually doing?

Reply:
In those tables, each row contains the data for multiple traits, so we need to return multiple values per row. This yield return is just a way of turning a single row into a collection of entities.

Reply:
Then it definitely needs a good comment - if I have to lookup stack overflow to work out what it's doing then it's not good.
A lot of this generic code is really clean - but it's also getting difficult to work out what it's doing - in general it needs more comments.

Reply:
Agreed, I've had 'improve the documentation' on the backburner for a while now, but it's getting to the point where I'm going to need to sit down for a couple of hours and make sure all the newer code is better documented, as per #8

Reply:
Which is the bad habit you need to avoid - get into the good habit of commenting as you go.

Reply:
I added some comments to this section in 23da668, will aim to be more thorough going forward.